### ISSUE TITLE
Why does firestore / AngularFire SDK packs multiple promises to one request?

### ISSUE THREAD

**Issue by jimmykane** [firebase/firebase-js-sdk#1447]
### Version info

**Angular:**  7.1.4

**Firebase:**5.7.1 

**AngularFire:**5.1.1

**Other (e.g. Ionic/Cordova, Node, browser, operating system):**

### How to reproduce these conditions

**Failing test unit, Plunkr, or JSFiddle demonstrating the problem**

**Steps to set up and reproduce**

Enforce Slow network

**Sample data and security rules**

<-- include/attach/link to some json sample data (or provide credentials to a sanitized, test Firebase project) -->

### Debug output

** Errors in the JavaScript console **

** Output from `firebase.database().enableLogging(true);` **

** Screenshots **

### Expected behavior

Many requests for multiple promises 

### Actual behavior

One request for multiple promises and that is problematic 

Here is how I write a document and it's subcollections:

```
public async setEvent(event: EventInterface): Promise<void[]> {
    return new Promise<void[]>(async (resolve, reject) => {
      const writePromises: Promise<void>[] = [];
      event.setID(event.getID() || this.afs.createId());
      event.getActivities()
        .forEach((activity) => {
          activity.setID(activity.getID() || this.afs.createId());
          writePromises.push(this.afs.collection('events').doc(event.getID()).collection('activities').doc(activity.getID()).set(activity.toJSON()));
          activity.getAllStreams().forEach((stream) => {
            this.logger.info(`Steam ${stream.type} has size of GZIP ${getSize(this.getBlobFromStreamData(stream.data))}`);
            writePromises.push(this.afs
              .collection('events')
              .doc(event.getID())
              .collection('activities')
              .doc(activity.getID())
              .collection('streams')
              .doc(stream.type) // @todo check this how it behaves
              .set({
                type: stream.type,
                data: this.getBlobFromStreamData(stream.data),
              }))
          });
        });
      try {
        await Promise.all(writePromises);
        await this.afs.collection('events').doc(event.getID()).set(event.toJSON());
        resolve()
      } catch (e) {
        Raven.captureException(e);
        // Try to delete the parent entity and all subdata
        await this.deleteEvent(event.getID());
        reject('Something went wrong')
      }
    })
  }
``` 

However when I look at the network tab:

![image](https://user-images.githubusercontent.com/1268723/50444433-c7eb8480-0908-11e9-8536-1f64b570024e.png)

I see one request firing up, well ok so far , req_0 data is my activity but looking further on the same request I can see:

enter image description here
![image](https://user-images.githubusercontent.com/1268723/50444436-cae67500-0908-11e9-854b-ed28891a3e7e.png)

So it adds more data and that should not happen because: a) I pass the size of the request to the firestore (1mb) b) due to slow connection I pass the time limit to write.

Most interesting is that this behavior happens when I have a slow network.

Here is the payload of the request example:

![image](https://user-images.githubusercontent.com/1268723/50444460-e18ccc00-0908-11e9-99fc-fa8c8b7bfbfc.png)

I am not sure if this angular fire or firebase but this is completly unwanted behavior

**Comment by mikelehen** [comment]
Firestore internally creates a bidirectional network channel to the backend using a Google library called [webchannel](https://github.com/google/closure-library/blob/e52d9b5c35dbbd93c73b06b44e38ed061e8ae45c/closure/goog/labs/net/webchannel.js#L16).  We send messages over this channel to the backend, and webchannel may batch them together when making individual requests to the backend.

Can you elaborate on why this is causing a problem for you?

**Comment by jimmykane** [comment]
@mikelehen I can explain sure. 

I am developing a web app that allows a user to upload a GPX file. That file is parsed from the JS code and send as documents for a collection. 
That document has a subcollections called streams that actually hold things latitude and longitude data or heartrate data. 

Those documents are about 250kib size. 

Upon a file processing it can have from 2 to 10 of those streams. 

10 * 250kib is about 2,5mib 

So far all works well whith a healthy internet connection. 

However, here comes the problem , if the internet connection is not fast for uploads on the client , those 10X2.5 requests are packed, and they pass the upload time limits as you understand, thus making it impossible to upload ALL at once. 

I hoped that each of those requests could succeed if they are isolated but packing them to one makes this quite heavy and out of scope to use promises. 

Correct me if I am wrong, but when I create multiple promises I do expect the benefit of parallel requests no? 

Does this make any sense at all?

**Comment by mikelehen** [comment]
@jimmykane Thanks, this is helpful.  

Can you explain what the "upload time limits" you're running into are?  Are you referring to some sort of browser limit or something?

As for parallel requests, I don't think that would help anything, and it might actually hurt.  You're fundamentally limited by your upload bandwidth and if the SDK did multiple requests at once, that would actually incur extra overhead (each request has request / response headers that would be repeated, etc.) and potentially make it actually a little slower.  So if your goal is to increase speed, I don't think separate requests will help.

If you wanted to try the experiment, you could try doing your writes using Firestore transactions instead of plain set() calls.  For various reasons, transactions don't actually make use of the WebChannel bidirectional channel, and so you would likely see parallel requests to the backend.  So you could see if that makes a difference.

**Comment by jimmykane** [comment]
I will need to retest with transactions  but I think regarding that I was hitting when I used a transaction instead of promises. I don't like promises so much to be honest as its not atomic. 
> Maximum API request size | 10 MiB


To be honest in the current setup I dont know what limit I am hitting. There is no response from the server just a failed request and the catch part of the code is never triggered. Looks like it never completes that promise all call. 

[Here is a video](https://drive.google.com/file/d/1z1tOyDN-1nnZsznvR3GRr-k33f6x2TQm/view), and btw if you want to test this live https://quantified-self-io.appspot.com provided that you have a big .fit file or I can provide the data etc. 

Here is a video link, you can stream or download. 

What you can see is that: 
- I have video streaming at home thus my connection is slow -> There is only one request firing to save all those promises -> Never completes and I never get the catch call triggered. There is no response error from the server nor the SDK. 
- I turn off video streaming -> all works fine as you can see and split in many requests in parallel. 

In short: if I do the above inside a transaction either slow network or not my request never completes. 

Now (as you might have noticed in the console) these are the "streams" that are inside the promise in the above code :


```
 EventService   Steam Latitude has size of GZIP 51.738 KiB 
console.js:35  EventService   Steam Longitude has size of GZIP 52.309 KiB 
console.js:35  EventService   Steam Heart Rate has size of GZIP 53.830 KiB 
console.js:35  EventService   Steam Distance has size of GZIP 283.961 KiB 
console.js:35  EventService   Steam Absolute Pressure has size of GZIP 257.572 KiB 
console.js:35  EventService   Steam Sea Level Pressure has size of GZIP 19.088 KiB 
console.js:35  EventService   Steam Altitude has size of GZIP 185.035 KiB 
console.js:35  EventService   Steam Cadence has size of GZIP 134.574 KiB 
console.js:35  EventService   Steam Power has size of GZIP 134.322 KiB 
console.js:35  EventService   Steam Speed has size of GZIP 66.816 KiB 
console.js:35  EventService   Steam Pace has size of GZIP 100.584 KiB 
console.js:35  EventService   Steam Temperature has size of GZIP 186.092 KiB 
console.js:35  EventService   Steam Vertical Speed has size of GZIP 38.418 KiB 
console.js:35  EventService   Steam EHPE has size of GZIP 9.641 KiB 
console.js:35  EventService   Steam EVPE has size of GZIP 10.002 KiB 
console.js:35  EventService   Steam Number of Satellites has size of GZIP 6.385 KiB 
console.js:35  EventService   Steam Satellite 5 Best SNR has size of GZIP 11.500 KiB 
console.js:35  EventService   Steam Battery Charge has size of GZIP 2.229 KiB 
console.js:35  EventService   Steam Battery Current has size of GZIP 63.063 KiB 
console.js:35  EventService   Steam Battery Voltage has size of GZIP 73.996 KiB 
console.js:35  EventService   Steam Heart Rate has size of GZIP 1.377 KiB 
console.js:35  EventService   Steam IBI has size of GZIP 336.229 KiB 
console.js:35  EventService   Steam Distance has size of GZIP 70 bytes 
console.js:35  EventService   Steam Absolute Pressure has size of GZIP 116 bytes 
console.js:35  EventService   Steam Sea Level Pressure has size of GZIP 72 bytes 
console.js:35  EventService   Steam Altitude has size of GZIP 86 bytes 
console.js:35  EventService   Steam Cadence has size of GZIP 78 bytes 
console.js:35  EventService   Steam Power has size of GZIP 58 bytes 
console.js:35  EventService   Steam Speed has size of GZIP 58 bytes 
console.js:35  EventService   Steam Pace has size of GZIP 58 bytes 
console.js:35  EventService   Steam Temperature has size of GZIP 192 bytes 
console.js:35  EventService   Steam Vertical Speed has size of GZIP 76 bytes 
console.js:35  EventService   Steam Battery Charge has size of GZIP 64 bytes 
console.js:35  EventService   Steam Battery Current has size of GZIP 104 bytes 
console.js:35  EventService   Steam Battery Voltage has size of GZIP 96 bytes 
console.js:35  EventService   Steam IBI has size of GZIP 44 bytes 
console.js:35  EventService   Steam Latitude has size of GZIP 49.699 KiB 
console.js:35  EventService   Steam Longitude has size of GZIP 52.348 KiB 
console.js:35  EventService   Steam Distance has size of GZIP 60.047 KiB 
console.js:35  EventService   Steam Absolute Pressure has size of GZIP 80.139 KiB 
console.js:35  EventService   Steam Sea Level Pressure has size of GZIP 4.752 KiB 
console.js:35  EventService   Steam Altitude GPS has size of GZIP 32.824 KiB 
console.js:35  EventService   Steam Altitude has size of GZIP 60.895 KiB 
console.js:35  EventService   Steam Cadence has size of GZIP 41.895 KiB 
console.js:35  EventService   Steam Power has size of GZIP 214 bytes 
console.js:35  EventService   Steam Speed has size of GZIP 19.463 KiB 
console.js:35  EventService   Steam Pace has size of GZIP 29.393 KiB 
console.js:35  EventService   Steam Temperature has size of GZIP 59.086 KiB 
console.js:35  EventService   Steam Vertical Speed has size of GZIP 11.889 KiB 
console.js:35  EventService   Steam EHPE has size of GZIP 6.061 KiB 
console.js:35  EventService   Steam EVPE has size of GZIP 6.711 KiB 
console.js:35  EventService   Steam Number of Satellites has size of GZIP 1.314 KiB 
console.js:35  EventService   Steam Satellite 5 Best SNR has size of GZIP 11.016 KiB 
console.js:35  EventService   Steam Battery Charge has size of GZIP 788 bytes 
console.js:35  EventService   Steam Battery Current has size of GZIP 18.729 KiB 
console.js:35  EventService   Steam Battery Voltage has size of GZIP 22.006 KiB 
console.js:35  EventService   Steam IBI has size of GZIP 44 bytes 
```

As you can see they are not big writes but all combined in a Transaction it can fail. 

However, somehow I get the feeling that when a slow network is detected some SDK magic wraps it all to a transactional kinda way or in any case, firestore is unresponsive or unusable. 

Please watch the video carefully and thanks for taking the time to look into this. Something is funky with the SDK. 


PS. I don't have persistence enabled on my app during the above and have tested that , having the same result and worse. (worse because if those requests don't finish there is no sync with the online DB) 


| am really lost here :-(

**Comment by mikelehen** [comment]
Thanks for the video and extra details.  I think I was able to reproduce the issue.  I used the 362kb multi-sport-with-heartrate.fit sample file from http://developer.garmin.com/garmin-connect-api/sample-data/ and used the Network Link Conditioner on my mac to simulate a 200kbps network connection.  It eventually succeeded, but there were many cancelled webchannel requests along the way.  When I used 100kbps, it didn't complete after ~10 minutes and didn't seem to be making progress.

I can try to isolate the issue a bit more and then talk to somebody more familiar with the webchannel transport to see what might be going on and if there's something we can do to fix it.

Until then, you could try using transactions like I mentioned before, or else wait for each Promise to complete before starting the next upload (rather than starting them all and using Promise.all()).

**Comment by mikelehen** [comment]
By the way, as a general consideration, if you wanted you could use Firebase Storage to store the streams since they seem to just be binary data.  It's fine to store them in Firestore, but Firebase Storage would likely be faster / cheaper since it's optimized for binary data.  Up to you though.

**Comment by mikelehen** [comment]
Also, would it be possible for you to run a http://www.speedtest.net/ in both the "good" (fast internet) state and the "bad" (slow internet) state.  I'd like to get a better idea of under what conditions you end up having trouble.  Thanks!

**Comment by jimmykane** [comment]
Great stuff @mikelehen ! 

I am somehow relieved we have spotted an issue or some consideration. 

Regarding the  Storage option, for now, this could do as if I need storage I need to manage another service / get / read / write /sync state. Call me lazy but for huge amounts of data 50(mb json files), that boils down to ~5mb reads from firestore.. I know it will hit a limit eventually but as my step one its my way to go. 

I would like if the firestore team allowed users to write big blobs as a special document entity that we could use natively within the sdk  and not having to manage another permission related store, implement requests etc. 

With all my honesty it's not production but rather experiment / hobby project.

**Comment by mikelehen** [comment]
Just as an update, we've identified the timeout that's causing an issue here and are investigating increasing it significantly so that hopefully it no longer comes into play.  Are you blocked by this issue, or can you wait some time for us to get a potential fix out?

**Comment by jimmykane** [comment]
Hi @mikelehen  and sorry for the late reply. 
No I am not blocked in regards to this issue and I am happy you have increased the timeouts. 

If any help is needed from my side (as of testing) feel free to ask. 

Keep up the hard work, have a great day and a great start of the week.

**Comment by rommelpe** [comment]
@jimmykane haven’t heard from you in a while. I would like to check if you are still facing this issue. 

You may try upgrading your JS SDK to the [latest version](https://firebase.google.com/support/release-notes/js) (v7.2.1) and let us know how it goes.

**Comment by jimmykane** [comment]
Hey that is correct. I have not faced this issue anymore.  Imo feel free to
close this.
